File: frontend/frontend/src/services/bookingService.js
import axiosInstance from "../api/axiosConfig";

/* ==========================
   GET ALL BOOKINGS (ADMIN)
========================== */
export const getAllBookings = async () => {
  const response = await axiosInstance.get("/api/bookings");
  return response.data;
};

/* ==========================
   GET BOOKINGS BY USER
========================== */
export const getBookingsByUser = async (userId) => {
  const response = await axiosInstance.get(
    `/api/bookings/user/${userId}`
  );
  return response.data;
};

/* ==========================
   CREATE BOOKING
========================== */
export const createBooking = async (bookingData) => {
  const response = await axiosInstance.post(
    "/api/bookings",
    bookingData
  );
  return response.data;
};

/* ==========================
   UPDATE BOOKING
========================== */
export const updateBooking = async (id, bookingData) => {
  const response = await axiosInstance.put(
    `/api/bookings/${id}`,
    bookingData
  );
  return response.data;
};

/* ==========================
   CANCEL BOOKING
========================== */
export const cancelBooking = async (id) => {
  const response = await axiosInstance.put(
    `/api/bookings/${id}/cancel`
  );
  return response.data;
};

/* ==========================
   DELETE BOOKING
========================== */
export const deleteBooking = async (id) => {
  const response = await axiosInstance.delete(
    `/api/bookings/${id}`
  );
  return response.data;
};

File: backend/ticket-booking-backend/ticket-booking-backend/src/main/java/com/example/ticket_booking_backend/service/BookingService.java
package com.example.ticket_booking_backend.service;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.example.ticket_booking_backend.model.Booking;
import com.example.ticket_booking_backend.model.Event;
import com.example.ticket_booking_backend.repository.BookingRepository;
import com.example.ticket_booking_backend.repository.EventRepository;

@Service
public class BookingService {

    @Autowired
    private BookingRepository bookingRepository;

    @Autowired
    private EventRepository eventRepository;

    /* ==========================
       CREATE BOOKING
    ========================== */

    public Booking saveBooking(Booking booking) {

        Event event = eventRepository.findById(booking.getEventId())
                .orElseThrow(() ->
                        new RuntimeException("Event not found"));

        // Check seat availability
        if (booking.getNumberOfTickets() > event.getAvailableSeats()) {
            throw new RuntimeException("Not enough seats available.");
        }

        // Calculate Total Amount
        booking.setTotalAmount(
                booking.getNumberOfTickets() * event.getTicketPrice()
        );

        // Reduce Available Seats
        event.setAvailableSeats(
                event.getAvailableSeats() - booking.getNumberOfTickets()
        );

        eventRepository.save(event);

        // Default Status
        if (booking.getStatus() == null || booking.getStatus().isBlank()) {
            booking.setStatus("CONFIRMED");
        }

        return bookingRepository.save(booking);
    }

    /* ==========================
       GET ALL BOOKINGS (ADMIN)
    ========================== */

    public List<Booking> getAllBookings() {
        return bookingRepository.findAll();
    }

    /* ==========================
       GET BOOKINGS BY USER
    ========================== */

    public List<Booking> getBookingsByUser(Long userId) {
        return bookingRepository.findByUserId(userId);
    }

    /* ==========================
       DELETE BOOKING
    ========================== */

    public void deleteBooking(Long id) {
        bookingRepository.deleteById(id);
    }

    /* ==========================
       UPDATE BOOKING
    ========================== */

    public Booking updateBooking(Long id, Booking updatedBooking) {

        Booking existingBooking = bookingRepository.findById(id)
                .orElseThrow(() ->
                        new RuntimeException("Booking not found"));

        existingBooking.setUserId(updatedBooking.getUserId());
        existingBooking.setEventId(updatedBooking.getEventId());
        existingBooking.setNumberOfTickets(updatedBooking.getNumberOfTickets());
        existingBooking.setTotalAmount(updatedBooking.getTotalAmount());
        existingBooking.setStatus(updatedBooking.getStatus());

        return bookingRepository.save(existingBooking);
    }

    /* ==========================
       CANCEL BOOKING
    ========================== */

    public Booking cancelBooking(Long id) {

        Booking booking = bookingRepository.findById(id)
                .orElseThrow(() ->
                        new RuntimeException("Booking not found"));

        if ("CANCELLED".equalsIgnoreCase(booking.getStatus())) {
            return booking;
        }

        Event event = eventRepository.findById(booking.getEventId())
                .orElseThrow(() ->
                        new RuntimeException("Event not found"));

        // Restore Seats
        event.setAvailableSeats(
                event.getAvailableSeats() + booking.getNumberOfTickets()
        );

        eventRepository.save(event);

        booking.setStatus("CANCELLED");

        return bookingRepository.save(booking);
    }
}